File: c++/p8g.cpp
#include "p8g.hpp"

#include <memory>

int p8g::width, p8g::height;
int p8g::keyCode;
bool p8g::keyIsPressed;
float p8g::mouseX, p8g::mouseY;
int p8g::mouseButton;
bool p8g::mouseIsPressed;
float p8g::deltaTime;

void p8g::applyMatrix(float a, float b, float c, float d, float e, float f) {
    p8g_apply_matrix(a, b, c, d, e, f);
}

void p8g::background(float gray) {
    float background[] = { gray, gray, gray, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_background(background);
}

void p8g::background(float gray, float alpha) {
    float background[] = { gray, gray, gray, alpha };
    p8g_background(background);
}

void p8g::background(float v1, float v2, float v3) {
    float background[] = { v1, v2, v3, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_background(background);
}

void p8g::background(float v1, float v2, float v3, float alpha) {
    float background[] = { v1, v2, v3, alpha };
    p8g_background(background);
}

void p8g::background(float color[4]) {
    p8g_background(color);
}

void p8g::colorMode(int mode) {
    p8g_color_mode((p8g_color_mode_t) mode);
}

void p8g::ellipse(float x, float y, float w, float h) {
    p8g_ellipse(x, y, w, h);
}

void p8g::ellipseMode(int mode) {
    p8g_ellipse_mode((p8g_ellipse_mode_t) mode);
}

void p8g::fill(float gray) {
    float fill[] = { gray, gray, gray, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_fill(fill);
}

void p8g::fill(float gray, float alpha) {
    float fill[] = { gray, gray, gray, alpha };
    p8g_fill(fill);
}

void p8g::fill(float v1, float v2, float v3) {
    float fill[] = { v1, v2, v3, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_fill(fill);
}

void p8g::fill(float v1, float v2, float v3, float alpha) {
    float fill[] = { v1, v2, v3, alpha };
    p8g_fill(fill);
}

void p8g::fill(float color[4]) {
    p8g_fill(color);
}

void p8g::image(p8g::Image img, float x, float y) {
    p8g_image(img, x, y, img.width, img.height, 0, 0, img.width, img.height);
}

void p8g::image(p8g::Image img, float x, float y, float w) {
    p8g_image(img, x, y, w, img.height, 0, 0, img.width, img.height);
}

void p8g::image(p8g::Image img, float x, float y, float w, float h) {
    p8g_image(img, x, y, w, h, 0, 0, img.width, img.height);
}

void p8g::image(p8g::Image img, float dx, float dy, float dw, float dh, float sx) {
    p8g_image(img, dx, dy, dw, dh, sx, 0, img.width, img.height);
}

void p8g::image(p8g::Image img, float dx, float dy, float dw, float dh, float sx, float sy) {
    p8g_image(img, dx, dy, dw, dh, sx, sy, img.width, img.height);
}

void p8g::image(p8g::Image img, float dx, float dy, float dw, float dh, float sx, float sy, float sw) {
    p8g_image(img, dx, dy, dw, dh, sx, sy, sw, img.height);
}

void p8g::image(p8g::Image img, float dx, float dy, float dw, float dh, float sx, float sy, float sw, float sh) {
    p8g_image(img, dx, dy, dw, dh, sx, sy, sw, sh);
}

void p8g::imageMode(int mode) {
    p8g_image_mode((p8g_image_mode_t) mode);
}

void p8g::line(float x1, float y1, float x2, float y2) {
    p8g_line(x1, y1, x2, y2);
}

p8g::Image p8g::loadImage(std::string filename) {
    return p8g_load_image(filename.c_str());
}

p8g::Font p8g::loadFont(std::string filename) {
    return p8g_load_font(filename.c_str());
}

int p8g::millis() {
    return p8g_time() * 1000.f;
}

void p8g::noFill() {
    p8g_no_fill();
}

void p8g::noSmooth() {
    p8g_no_smooth();
}

void p8g::noStroke() {
    p8g_no_stroke();
}

void p8g::noTint() {
    p8g_no_tint();
}

void p8g::point(float x, float y) {
    p8g_point(x, y);
}

void p8g::pop() {
    p8g_pop();
}

void p8g::push() {
    p8g_push();
}

float p8g::random() {
    return p8g_random(0.f, 1.f);
}

float p8g::random(float max) {
    return p8g_random(0.f, max);
}

float p8g::random(float min, float max) {
    return p8g_random(min, max);
}

void p8g::randomSeed(int seed) {
    p8g_random_seed(seed);
}

void p8g::rect(float x, float y, float w, float h) {
    p8g_rect(x, y, w, h);
}

void p8g::rectMode(int mode) {
    p8g_rect_mode((p8g_rect_mode_t) mode);
}

void p8g::resetMatrix() {
    p8g_reset_matrix();
}

void p8g::rotate(float angle) {
    p8g_rotate(angle);
}

static struct {
    void (*draw)(void);
    void (*keyPressed)(void);
    void (*keyReleased)(void);
    void (*mouseMoved)(void);
    void (*mousePressed)(void);
    void (*mouseReleased)(void);
    void (*mouseWheel)(float);
} captures;
void p8g::_run(p8g::RunArgs runArgs, void (*draw)(void), void (*keyPressed)(void), void (*keyReleased)(void), void (*mouseMoved)(void), void (*mousePressed)(void), void (*mouseReleased)(void), void (*mouseWheel)(float)) {
    p8g::width = runArgs.width;
    p8g::height = runArgs.height;
    captures.draw = draw;
    captures.keyPressed = keyPressed;
    captures.keyReleased = keyReleased;
    captures.mouseMoved = mouseMoved;
    captures.mousePressed = mousePressed;
    captures.mouseReleased = mouseReleased;
    captures.mouseWheel = mouseWheel;
    p8g_run((p8g_sketch_t) {
        .width = runArgs.width,
        .height = runArgs.height,
        .title = runArgs.title.c_str(),
        .full_screen = runArgs.fullScreen,
        .draw = [](float deltaTime) {
            p8g::deltaTime = deltaTime * 1000.f;
            captures.draw();
        },
        .key_pressed = [](int keyCode) {
            p8g::keyCode = keyCode;
            p8g::keyIsPressed = true;
            captures.keyPressed();
        },
        .key_released = [](int keyCode) {
            p8g::keyCode = keyCode;
            p8g::keyIsPressed = false;
            captures.keyReleased();
        },
        .mouse_moved = [](float mouseX, float mouseY) {
            p8g::mouseX = mouseX;
            p8g::mouseY = mouseY;
            captures.mouseMoved();
        },
        .mouse_pressed = [](int mouseButton) {
            p8g::mouseButton = mouseButton;
            p8g::mouseIsPressed = true;
            captures.mousePressed();
        },
        .mouse_released = [](int mouseButton) {
            p8g::mouseButton = mouseButton;
            p8g::mouseIsPressed = false;
            captures.mouseReleased();
        },
        .mouse_wheel = [](float deltaX, float deltaY) {
            captures.mouseWheel(deltaY);
        },
    });
}

void p8g::scale(float s) {
    p8g_scale(s, s);
}

void p8g::scale(float x, float y) {
    p8g_scale(x, y);
}

void p8g::smooth() {
    p8g_smooth();
}

void p8g::stroke(float gray) {
    float stroke[] = { gray, gray, gray, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_stroke(stroke);
}

void p8g::stroke(float gray, float alpha) {
    float stroke[] = { gray, gray, gray, alpha };
    p8g_stroke(stroke);
}

void p8g::stroke(float v1, float v2, float v3) {
    float stroke[] = { v1, v2, v3, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_stroke(stroke);
}

void p8g::stroke(float v1, float v2, float v3, float alpha) {
    float stroke[] = { v1, v2, v3, alpha };
    p8g_stroke(stroke);
}

void p8g::stroke(float color[4]) {
    p8g_stroke(color);
}

void p8g::strokeWeight(float weight) {
    p8g_stroke_weight(weight);
}

void p8g::text(std::string str, float x, float y) {
    p8g_text(str.c_str(), x, y);
}

void p8g::textFont(p8g::Font font) {
    p8g_text_font(font);
}

void p8g::textSize(float size) {
    p8g_text_size(size);
}

void p8g::tint(float gray) {
    float tint[] = { gray, gray, gray, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_tint(tint);
}

void p8g::tint(float gray, float alpha) {
    float tint[] = { gray, gray, gray, alpha };
    p8g_tint(tint);
}

void p8g::tint(float v1, float v2, float v3) {
    float tint[] = { v1, v2, v3, p8g_peek_color_mode() == p8g::RGB ? 255.f : 1.f };
    p8g_tint(tint);
}

void p8g::tint(float v1, float v2, float v3, float alpha) {
    float tint[] = { v1, v2, v3, alpha };
    p8g_tint(tint);
}

void p8g::tint(float color[4]) {
    p8g_tint(color);
}

void p8g::translate(float x, float y) {
    p8g_translate(x, y);
}

void p8g::triangle(float x1, float y1, float x2, float y2, float x3, float y3) {
    p8g_triangle(x1, y1, x2, y2, x3, y3);
}

File: js/src/p8g.js
import ModuleFactory from '../build/p8g.js';
import p8gWasm from '../../c/p8g.wasm';
const Module = await ModuleFactory({
  canvas: (() => {
    const style = document.createElement('style');
    document.head.appendChild(style);
    style.sheet.insertRule(`
      canvas.p8g {
        border: 0px none;
        background-color: black;
        image-rendering: optimizeSpeed;
        image-rendering: -moz-crisp-edges;
        image-rendering: -o-crisp-edges;
        image-rendering: -webkit-optimize-contrast;
        image-rendering: optimize-contrast;
        image-rendering: crisp-edges;
        image-rendering: pixelated;
        -ms-interpolation-mode: nearest-neighbor;
        margin: -0.375px 0 0 -0.375px;
      }
    `);
    style.sheet.insertRule(`
      .hidden {
        display: none;
      }
    `);
    const canvasWrapper = document.createElement('div');
    document.body.appendChild(canvasWrapper);
    canvasWrapper.style.cssText = 'display: inline-flex; overflow: hidden;';
    const canvas = document.createElement('canvas');
    canvasWrapper.appendChild(canvas);
    canvas.classList.add('p8g');
    canvas.classList.add('hidden');
    canvas.addEventListener('contextmenu', (e) => {
      e.preventDefault();
    });
    canvas.tabIndex = -1;
    canvas.getContext = ((getContext) => {
      return function (contextType, contextAttributes) {
        if (contextAttributes) {
          contextAttributes.antialias = false;
          contextAttributes.imageSmoothingEnabled = false;
          contextAttributes.preserveDrawingBuffer = true;
        }
        return getContext.call(this, contextType, contextAttributes);
      };
    })(canvas.getContext);
    return canvas;
  })(),
  locateFile: (path, scriptDirectory) => {
    if (path.endsWith('.wasm')) {
      return 'data:application/octet-stream;base64,' + p8gWasm;
    }
    return scriptDirectory + path;
  },
});
/**
 * @type {{
 * draw: () => void,
 * keyPressed: () => void,
 * keyReleased: () => void,
 * mouseMoved: () => void,
 * mousePressed: () => void,
 * mouseReleased: () => void,
 * mouseWheel: (deltaY: number) => void
 * }}
 */
const p8g = {
  draw: () => {},
  keyPressed: () => {},
  keyReleased: () => {},
  mouseMoved: () => {},
  mousePressed: () => {},
  mouseReleased: () => {},
  mouseWheel: (deltaY) => {},
};
export default p8g;
export const CORNER = 0;
export const CORNERS = 1;
export const RADIUS = 2;
export const CENTER = 3;
export const RGB = 0;
export const HSB = 1;
export const HSL = 2;
export let width = 100;
export let height = 100;
export let keyCode = 0;
export let keyIsPressed = false;
export let mouseX = 0;
export let mouseY = 0;
export let mouseButton = 0;
export let mouseIsPressed = false;
export let deltaTime = 0;
/** @type {(a: number, b: number, c: number, d: number, e: number, f: number) => void} */
export const applyMatrix = Module.cwrap('p8g_apply_matrix', 'void', [
  'number',
  'number',
  'number',
  'number',
]);
/**
 * @type {{
* (values: [number, number, number, number]) => void;
* (gray: number) => void;
* (gray: number, alpha: number) => void;
* (v1: number, v2: number, v3: number) => void;
* (v1: number, v2: number, v3: number, alpha: number) => void;
* }}
*/
export const background = (() => {
  const background = Module.cwrap('p8g_background', 'void', ['array']);
  const background1 = Module.cwrap('background1', 'void', ['number']);
  const background2 = Module.cwrap('background2', 'void', ['number', 'number']);
  const background3 = Module.cwrap('background3', 'void', [
    'number',
    'number',
    'number',
  ]);
  const background4 = Module.cwrap('background4', 'void', [
    'number',
    'number',
    'number',
    'number',
  ]);
  return (...args) => {
    switch (args.length) {
      case 1:
        if (Array.isArray(args[0])) {
          background(new Uint8Array(new Float32Array(args[0]).buffer));
        } else {
          background1(args[0]);
        }
        break;
      case 2:
        background2(args[0], args[1]);
        break;
      case 3:
        background3(args[0], args[1], args[2]);
        break;
      case 4:
        background4(args[0], args[1], args[2], args[3]);
        break;
    }
  };
})();
/** @type {(mode: RGB | HSB | HSL) => void} */
export const colorMode = Module.cwrap('p8g_color_mode', 'void', ['number']);
/** @type {(x: number, y: number, w: number, h: number) => void} */
export const ellipse = Module.cwrap('p8g_ellipse', 'void', [
  'number',
  'number',
  'number',
  'number',
]);
/** @type {(mode: CORNER | CORNERS | RADIUS | CENTER) => void} */
export const ellipseMode = Module.cwrap('p8g_ellipse_mode', 'void', ['number']);
/**
 * @type {{
* (values: [number, number, number, number]) => void;
* (gray: number) => void;
* (gray: number, alpha: number) => void;
* (v1: number, v2: number, v3: number) => void;
* (v1: number, v2: number, v3: number, alpha: number) => void;
* }}
*/
export const fill = (() => {
  const fill = Module.cwrap('p8g_fill', 'void', ['array']);
  const fill1 = Module.cwrap('fill1', 'void', ['number']);
  const fill2 = Module.cwrap('fill2', 'void', ['number', 'number']);
  const fill3 = Module.cwrap('fill3', 'void', ['number', 'number', 'number']);
  const fill4 = Module.cwrap('fill4', 'void', [
    'number',
    'number',
    'number',
    'number',
  ]);
  return (...args) => {
    switch (args.length) {
      case 1:
        if (Array.isArray(args[0])) {
          fill(new Uint8Array(new Float32Array(args[0]).buffer));
        } else {
          fill1(args[0]);
        }
        break;
      case 2:
        fill2(args[0], args[1]);
        break;
      case 3:
        fill3(args[0], args[1], args[2]);
        break;
      case 4:
        fill4(args[0], args[1], args[2], args[3]);
        break;
    }
  };
})();
/** @typedef {{ _index: number, width: number, height: number }} Image */
/**
 * @type {{
 * (img: Image, x: number, y: number) => void;
 * (img: Image, x: number, y: number, width: number) => void;
 * (img: Image, x: number, y: number, width: number, height: number) => void;
 * (img: Image, dx: number, dy: number, dw: number, dh: number, sx: number) => void;
 * (img: Image, dx: number, dy: number, dw: number, dh: number, sx: number, sy: number) => void;
 * (img: Image, dx: number, dy: number, dw: number, dh: number, sx: number, sy: number, sw: number) => void;
 * (img: Image, dx: number, dy: number, dw: number, dh: number, sx: number, sy: number, sw: number, sh: number) => void;
 * }}
 */
export const image = (() => {
  const image = Module.cwrap('image', 'void', [
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
  ]);
  return (...args) => {
    if (args.length < 3 || !args[0]) {
      return;
    }
    switch (args.length) {
      case 3:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[0].width,
          args[0].height,
          0,
          0,
          args[0].width,
          args[0].height,
        );
        break;
      case 4:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[3],
          args[0].height,
          0,
          0,
          args[0].width,
          args[0].height,
        );
        break;
      case 5:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[3],
          args[4],
          0,
          0,
          args[0].width,
          args[0].height,
        );
        break;
      case 6:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[3],
          args[4],
          args[5],
          0,
          args[0].width,
          args[0].height,
        );
        break;
      case 7:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[3],
          args[4],
          args[5],
          args[6],
          args[0].width,
          args[0].height,
        );
        break;
      case 8:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[3],
          args[4],
          args[5],
          args[6],
          args[7],
          args[0].height,
        );
        break;
      case 9:
        image(
          args[0]._index,
          args[1],
          args[2],
          args[3],
          args[4],
          args[5],
          args[6],
          args[7],
          args[8],
        );
        break;
    }
  };
})();
/** @type {(mode: CORNER | CORNERS | RADIUS | CENTER) => void} */
export const imageMode = Module.cwrap('p8g_image_mode', 'void', ['number']);
/** @type {(x1: number, y1: number, x2: number, y2: number) => void} */
export const line = Module.cwrap('p8g_line', 'void', [
  'number',
  'number',
  'number',
  'number',
]);
/** @typedef {{ _index: number }} Font */
/** @type {(filename: string) => Promise<Font>} */
export const loadFont = (() => {
  const loadFont = Module.cwrap('loadFont', 'number', ['string']);
  return async (url) => {
    const data = await fetch(url)
      .then((res) => res.arrayBuffer())
      .then((arrayBuffer) => new Uint8Array(arrayBuffer));
    const { hostname, pathname } = new URL(url, document.baseURI);
    const path = hostname + pathname;
    const dir = path.slice(0, path.lastIndexOf('/'));
    Module.FS_createPath('/', dir, true, true);
    Module.FS_createDataFile('/', path, data, true, false);
    return {
      _index: loadFont(path),
    };
  };
})();
/** @type {(filename: string) => Promise<Image>} */
export const loadImage = (() => {
  const loadImage = Module.cwrap('loadImage', 'number', ['string']);
  return async (url) => {
    const data = await fetch(url)
      .then((res) => res.arrayBuffer())
      .then((arrayBuffer) => new Uint8Array(arrayBuffer));
    const { hostname, pathname } = new URL(url, document.baseURI);
    const path = hostname + pathname;
    const dir = path.slice(0, path.lastIndexOf('/'));
    Module.FS_createPath('/', dir, true, true);
    Module.FS_createDataFile('/', path, data, true, false);
    let i = loadImage(path);
    if (i === -1) {
      return {
        _index: -1,
        width: 0,
        height: 0,
      };
    }
    return {
      _index: (i >> 24) & 0xff,
      width: (i >> 12) & 0xfff,
      height: i & 0xfff,
    };
  };
})();
/** @type {() => number} */
export const millis = (() => {
  const time = Module.cwrap('p8g_time', 'float', []);
  return () => {
    return time() * 1000;
  };
})();
/** @type {() => void} */
export const noFill = Module.cwrap('p8g_no_fill', 'void', []);
/** @type {() => void} */
export const noSmooth = Module.cwrap('p8g_no_smooth', 'void', []);
/** @type {() => void} */
export const noStroke = Module.cwrap('p8g_no_stroke', 'void', []);
/** @type {() => void} */
export const noTint = Module.cwrap('p8g_no_tint', 'void', []);
/** @type {(x: number, y: number) => void} */
export const point = Module.cwrap('p8g_point', 'void', ['number', 'number']);
/** @type {() => void} */
export const pop = Module.cwrap('p8g_pop', 'void', []);
/** @type {() => void} */
export const push = Module.cwrap('p8g_push', 'void', []);
/**
 * @type {{
 * () => number;
 * (max: number) => number;
 * (min: number, max: number) => number;
 * }}
 */
export const random = (() => {
  const random = Module.cwrap('p8g_random', 'number', ['number', 'number']);
  return (...args) => {
    switch (args.length) {
      case 0:
        return random(0, 1);
      case 1:
        return random(0, args[0]);
      case 2:
        return random(args[0], args[1]);
    }
  };
})();
/** @type {(seed: number) => void} */
export const randomSeed = Module.cwrap('p8g_random_seed', 'void', ['number']);
/** @type {(x: number, y: number, w: number, h: number) => void} */
export const rect = Module.cwrap('p8g_rect', 'void', [
  'number',
  'number',
  'number',
  'number',
]);
/** @type {(mode: CORNER | CORNERS | RADIUS | CENTER) => void} */
export const rectMode = Module.cwrap('p8g_rect_mode', 'void', ['number']);
/** @type {() => void} */
export const resetMatrix = Module.cwrap('p8g_reset_matrix', 'void', []);
/** @type {(angle: number) => void} */
export const rotate = Module.cwrap('p8g_rotate', 'void', ['number']);

/** @type {(width: number, height: number) => HTMLCanvasElement} */
export const createCanvas = (() => {
  const run = Module.cwrap('run', 'void', [
    'number',
    'number',
    'string',
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
    'number',
  ]);
  const draw = (_deltaTime) => {
    deltaTime = _deltaTime * 1000;
    p8g.draw();
  };
  const keyPressed = (_keyCode) => {
    keyCode = _keyCode;
    keyIsPressed = true;
    p8g.keyPressed();
  };
  const keyReleased = (_keyCode) => {
    keyCode = _keyCode;
    keyIsPressed = false;
    p8g.keyReleased();
  };
  const mouseMoved = (_mouseX, _mouseY) => {
    mouseX = _mouseX;
    mouseY = _mouseY;
    p8g.mouseMoved();
  };
  const mousePressed = (_mouseButton) => {
    mouseButton = _mouseButton;
    mouseIsPressed = true;
    p8g.mousePressed();
  };
  const mouseReleased = (_mouseButton) => {
    mouseButton = _mouseButton;
    mouseIsPressed = false;
    p8g.mouseReleased();
  };
  const mouseWheel = (deltaX, deltaY) => {
    p8g.mouseWheel(deltaY);
  };
  return (_width, _height) => {
    width = _width;
    height = _height;
    Module.canvas.classList.remove('hidden');
    run(
      width,
      height,
      'Sketch',
      0,
      Module.addFunction(draw, 'vf'),
      Module.addFunction(keyPressed, 'vi'),
      Module.addFunction(keyReleased, 'vi'),
      Module.addFunction(mouseMoved, 'vff'),
      Module.addFunction(mousePressed, 'vi'),
      Module.addFunction(mouseReleased, 'vi'),
      Module.addFunction(mouseWheel, 'vff'),
    );
    return Module.canvas;
  };
})();
/**
 * @type {{
 * (s: number) => void;
 * (x: number, y: number) => void;
 * }}
 */
export const scale = (() => {
  const scale = Module.cwrap('p8g_scale', 'void', ['number']);
  return (...args) => {
    switch (args.length) {
      case 1:
        scale(args[0], args[0]);
        break;
      case 2:
        scale(args[0], args[1]);
        break;
    }
  };
})();
/** @type {() => void} */
export const smooth = Module.cwrap('p8g_smooth', 'void', []);
/**
 * @type {{
* (values: [number, number, number, number]) => void;
* (gray: number) => void;
* (gray: number, alpha: number) => void;
* (v1: number, v2: number, v3: number) => void;
* (v1: number, v2: number, v3: number, alpha: number) => void;
* }}
*/
export const stroke = (() => {
  const stroke = Module.cwrap('p8g_stroke', 'void', ['array']);
  const stroke1 = Module.cwrap('stroke1', 'void', ['number']);
  const stroke2 = Module.cwrap('stroke2', 'void', ['number', 'number']);
  const stroke3 = Module.cwrap('stroke3', 'void', [
    'number',
    'number',
    'number',
  ]);
  const stroke4 = Module.cwrap('stroke4', 'void', [
    'number',
    'number',
    'number',
    'number',
  ]);
  return (...args) => {
    switch (args.length) {
      case 1:
        if (Array.isArray(args[0])) {
          stroke(new Uint8Array(new Float32Array(args[0]).buffer));
        } else {
          stroke1(args[0]);
        }
        break;
      case 2:
        stroke2(args[0], args[1]);
        break;
      case 3:
        stroke3(args[0], args[1], args[2]);
        break;
      case 4:
        stroke4(args[0], args[1], args[2], args[3]);
        break;
    }
  };
})();
/** @type {(weight: number) => void} */
export const strokeWeight = Module.cwrap('p8g_stroke_weight', 'void', [
  'number',
]);
/** @type {(str: string, x: number, y: number) => void} */
export const text = Module.cwrap('p8g_text', 'void', [
  'string',
  'number',
  'number',
]);
/** @type {(font: Font) => void} */
export const textFont = (() => {
  const textFont = Module.cwrap('textFont', 'void', ['number']);
  return (font) => {
    textFont(font._index);
  };
})();
/** @type {(size: number) => void} */
export const textSize = Module.cwrap('p8g_text_size', 'void', ['number']);
/**
 * @type {{
 * (values: [number, number, number, number]) => void;
 * (gray: number) => void;
 * (gray: number, alpha: number) => void;
 * (v1: number, v2: number, v3: number) => void;
 * (v1: number, v2: number, v3: number, alpha: number) => void;
 * }}
 */
export const tint = (() => {
  const tint = Module.cwrap('p8g_tint', 'void', ['array']);
  const tint1 = Module.cwrap('tint1', 'void', ['number']);
  const tint2 = Module.cwrap('tint2', 'void', ['number', 'number']);
  const tint3 = Module.cwrap('tint3', 'void', ['number', 'number', 'number']);
  const tint4 = Module.cwrap('tint4', 'void', [
    'number',
    'number',
    'number',
    'number',
  ]);
  return (...args) => {
    switch (args.length) {
      case 1:
        if (Array.isArray(args[0])) {
          tint(new Uint8Array(new Float32Array(args[0]).buffer));
        } else {
          tint1(args[0]);
        }
        break;
      case 2:
        tint2(args[0], args[1]);
        break;
      case 3:
        tint3(args[0], args[1], args[2]);
        break;
      case 4:
        tint4(args[0], args[1], args[2], args[3]);
        break;
    }
  };
})();
/** @type {(x: number, y: number) => void} */
export const translate = Module.cwrap('p8g_translate', 'void', [
  'number',
  'number',
]);
/** @type {(x1: number, y1: number, x2: number, y2: number, x3: number, y3: number) => void} */
export const triangle = Module.cwrap('p8g_triangle', 'void', [
  'number',
  'number',
  'number',
  'number',
  'number',
  'number',
]);
